### PR TITLE
Validate deletion timestamp doesn't change on update

### PR DESCRIPTION
**Pull request by liggitt**
Fixes https://github.com/kubernetes/kubernetes/issues/24828

Updates validation to ensure that deletionTimestamp and deletionGracePeriodSeconds are not set during update operations to ensure those fields are only set during delete operations.

### CODE REVIEW

**Comment by liggitt** [comment]
@smarterclayton ptal regarding intended semantics on deletionTimestamp/deleteGracePeriodSeconds

@derekwaynecarr ptal at the namespace status update tests... were those just intended to test that the phase matched the presence/absence of a deletionTimestamp, or was mutating deletionTimestamp via a status update intended?

**Comment by derekwaynecarr** [comment]
@lavalamp @liggitt - I am +100 in favor of this change.  edit of a namespace should not have caused deletion timestamp to be set to trigger deletion of entire namespace!

**Comment by derekwaynecarr** [comment]
I would also like this to be cherry-picked.

**Comment by derekwaynecarr** [comment]
This was an oversight in the implementation.

See the initial design that only discussed DELETE triggering a change to DeletionTimestamp: https://github.com/kubernetes/kubernetes/blob/master/docs/design/namespaces.md#phases

>If a DELETE request occurs for a Namespace, the Namespace.ObjectMeta.DeletionTimestamp is set to the current server time. A namespace controller observes the change, and sets the Namespace.Status.Phase to Terminating.

It was NOT intended that update to a Namespace could trigger cascading deletion.

**Comment by zmerlynn** [comment]
FYI, the `milestone` @k8s-cherrypick-bot wants is the furthest back you want it cherrypicked to. I'm assuming v1.2, since right now we're not shipping v1.1s.

File: pkg/registry/namespace/etcd/etcd.go [hunk 1]
```diff
@@ -99,21 +101,67 @@ func (r *REST) Delete(ctx api.Context, name string, options *api.DeleteOptions)
...
+	// Ensure we have a UID precondition
+	if options == nil {
+		options = api.NewDeleteOptions(0)
+	}
+	if options.Preconditions == nil {
+		options.Preconditions = &api.Preconditions{}
+	}
+	if options.Preconditions.UID == nil {
+		options.Preconditions.UID = &namespace.UID
+	}
```
Comment (lavalamp) [inline pkg/registry/namespace/etcd/etcd.go]: Else, check that namespace.UID == options.Preconditions.UID? (i.e., bail early)

Comment (liggitt) [inline pkg/registry/namespace/etcd/etcd.go]: sure, do we return conflict errors in those cases?

File: pkg/registry/namespace/etcd/etcd.go [hunk 2]
```diff
@@ -99,21 +101,67 @@ func (r *REST) Delete(ctx api.Context, name string, options *api.DeleteOptions)
...
+			}
+			return nil, err
+		}
+
+		return out, nil
 	}
 
 	// prior to final deletion, we must ensure that finalizers is empty
 	if len(namespace.Spec.Finalizers) != 0 {
 		err = apierrors.NewConflict(api.Resource("namespaces"), namespace.Name, fmt.Errorf("The system is ensuring all content is removed from this namespace.  Upon completion, this namespace will automatically be purged by the system."))
```
Comment (lavalamp) [inline pkg/registry/namespace/etcd/etcd.go]: Does it get deleted automatically when the last finalizer is removed?

File: pkg/registry/namespace/etcd/etcd.go [hunk 3]
```diff
@@ -99,21 +101,67 @@ func (r *REST) Delete(ctx api.Context, name string, options *api.DeleteOptions)
 
 	namespace := nsObj.(*api.Namespace)
 
+	// Ensure we have a UID precondition
```
Comment (lavalamp) [inline pkg/registry/namespace/etcd/etcd.go]: These changes seem good but unrelated-- do they really need to be cherrypicked?

**Comment by lavalamp** [comment]
LGTM

File: pkg/registry/namespace/etcd/etcd.go (hunk 1 above)
Comment (lavalamp) [inline pkg/registry/namespace/etcd/etcd.go]: Yes, I think that's what @caesarxuchao did in the place I'm thinking of.

File: pkg/registry/namespace/etcd/etcd.go (hunk 2 above)
Comment (derekwaynecarr) [inline pkg/registry/namespace/etcd/etcd.go]: no, the namespace controller sends an additional delete call.

Comment (liggitt) [inline pkg/registry/namespace/etcd/etcd.go]: yes, the namespace controller handles that:
https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/namespace/namespace_controller_utils.go#L369-L376

Comment (lavalamp) [inline pkg/registry/namespace/etcd/etcd.go]: For garbage collector, we were planning on making the update path auto delete things with a deletion timestamp when the last finalizer is removed. Do you have any objections to changing that?

Comment (lavalamp) [inline pkg/registry/namespace/etcd/etcd.go]: That would work much better if multiple components were removing finalizers, assuming you can craft a reasonable PATCH request that removes a single finalizer.

File: pkg/registry/namespace/etcd/etcd.go (hunk 3 above)
Comment (liggitt) [inline pkg/registry/namespace/etcd/etcd.go]: UID preconditions didn't exist in 1.2, I'll adjust appropriately when I cherry pick

File: pkg/registry/namespace/etcd/etcd.go (hunk 1 above)
Comment (liggitt) [inline pkg/registry/namespace/etcd/etcd.go]: done

**Comment by liggitt** [comment]
comments addressed

**Comment by liggitt** [comment]
tagging per comment

**Pull request merged**